frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/EditFieldDrawer/edit-fieldtype/common/useEditFieldForm.ts

```diff
@@ -0,0 +1,93 @@
+import { useMemo } from 'react'
+import {
+  DeepPartial,
+  UnpackNestedValue,
+  useForm,
+  UseFormReturn,
+} from 'react-hook-form'
+import { useDebounce } from 'react-use'
+
+import { FieldBase } from '~shared/types/field'
+
+import { EditFieldProps } from './types'
+import { getButtonText } from './utils'
+
+type UseEditFieldFormProps<
+  FormShape,
+  FieldShape extends FieldBase,
+> = EditFieldProps<FieldShape> & {
+  transform: {
+    input: (field: FieldShape) => FormShape
+    output: (form: FormShape, originalField: FieldShape) => FieldShape
+  }
+}
+
+type UseEditFieldFormReturn<U> = UseFormReturn<U> & {
+  handleUpdateField: () => Promise<void>
+  buttonText: string
+  isSaveEnabled: boolean
+}
+
+export const useEditFieldForm = <FormShape, FieldShape extends FieldBase>({
+  field,
+  isPendingField,
+  handleChange,
+  handleSave,
+  transform,
+}: UseEditFieldFormProps<
+  FormShape,
+  FieldShape
+>): UseEditFieldFormReturn<FormShape> => {
+  const defaultValues = useMemo(
+    () => transform.input(field) as UnpackNestedValue<DeepPartial<FormShape>>,
+    [field, transform],
+  )
+  const editForm = useForm<FormShape>({
+    defaultValues,
+  })
+
+  const watchedInputs = editForm.watch()
+
+  useDebounce(
+    () => handleChange(transform.output(watchedInputs as FormShape, field)),
+    300,
+    Object.values(watchedInputs),
+  )
+
+  const handleUpdateField = editForm.handleSubmit((inputs) => {
+    const updatedFormField: FieldShape = transform.output(
+      inputs as FormShape,
```

Comment:
do you need to cast `FormShape` seeing how that type is already passed into `useForm` as a generic?

Same for the other casts

Reply:
yeah so `inputs` is of type `UnpackNestedValue<FormShape>`, not `FormShape`. I tried looking into what the `UnpackNestedValue` type does, but figured it was probably too much effort to try to understand. do you know what's going on here?

````typescript
export declare type NestedValue<TValue extends object = object> = {
    [$NestedValue]: never;
} & TValue;
export declare type UnpackNestedValue<T> = T extends NestedValue<infer U> ? U : T extends Date | FileList | File ? T : T extends object ? {
    [K in keyof T]: UnpackNestedValue<T[K]>;
} : T;
```

Reply:
This could be because of your defaultValues type in line 42. Any reason why it's casted to UnpackNestedValue?

Reply:
This could be because of your defaultValues type in line 42. Any reason why it's casted to UnpackNestedValue?

Reply:
yeah it's casted because the type required for `defaultValues` is also `UnpackNestedValue`

Reply:
managed to remove the casts by changing the types in `UseEditFieldFormProps`